<!DOCTYPE html>

<head>
  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta2/dist/css/bootstrap.min.css" rel="stylesheet"
    integrity="sha384-BmbxuPwQa2lc/FVzBcNJ7UAyJxM6wuqIj61tLrc4wSX0szH/Ev+nYRRuWlolflfl" crossorigin="anonymous">
  <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta2/dist/js/bootstrap.bundle.min.js"
    integrity="sha384-b5kHyXgcpbZJO/tY9Ul7kGkf1S0CWuKcCD38l8YkeH8z8QjE0GmW1gYU5S9FOnJ0" crossorigin="anonymous">
  </script>

  <script src="https://episphere.github.io/epibox/epibox.js"></script>
  <script src=https://cdnjs.cloudflare.com/ajax/libs/mathjs/3.3.0/math.min.js></script>
  <link href="style.css" rel="stylesheet">


  <h1>23andMe Risk Calculator</h1><br>
  <p>A polygenic risk score estimates the genetic risk of an individual for some disease or trait, calculated by aggregating the effect of many common variants associated with the condition. Read more 
    <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7612115/#:~:text=In%20the%20standard%20approach%20%5B5,powerful%20GWAS%20on%20the%20phenotype">here.</a>
  </p>
  <img src="prs2.png" alt="PRS equation" width="300" height="80"/><br>
  <li>beta = Weights (example is B as the per allele log odds ratio for the minor allele</li>
  <li>k = Allele dosage for the SNP (0,1, or 2</li>
  <li>n = number of SNPS</li><br>

  <b>Workflow for calculating risk:</b>
  <li>Retrieve scoring file from 
    <a href="https://www.pgscatalog.org/rest/">PGS catalog</a>
  </li>  
  <li>Retrieve genotype from 23 and me raw data file</li>
  <li>Match scoring file variants and calculate scores</li>
  <br>
</head>


<body>
  <!-- PGS scores------->
  <h2>PGS Risk Scores</h2>
  <input id="pgsInput" type="text" value="PGS000001" >
  <p id = "pgsFileInfo"></p>
  <button onclick="getPGSdata(), getPGSdata2()">enter</button>
  <textarea id="pgsArea" style="width:100%;color:lime;background-color:black;height:20em;font-size:small"></textarea>
  <br><br>
  
  <!-- Genomic data----->
  <h2>Genomic Data</h2>
  <input type="file" id="input-file" onchange="getFile(event, 'genomArea')">
  <textarea id="genomArea" class="content-target" style="width:100%;height:10em;font-size:small"></textarea>
  <br><br>

   <!-- Risk result------>
   <h2>Risk results</h2>
   <p id = "variantMatches"></p>
  <div id = resDiv style ="color: red; border-width:2px; 
  border-style:solid; border-color:#FF0000; padding: 0.7em;">Please load genomic file...</div>
  <br>

<div id = endDiv>
  <p class="info" style="font-size:10px;"> All computation takes place in this browser session. No data is transmited to a server, because there is no server supporting this application. The main purpose of this tool is indeed to illustrate how serverless statistical proceedures, particulalry data wrangling, 
    can migrate to where the data is without exposing the source data. </p>
</div>
  <script src="https://episphere.github.io/pgs/pgs.js"></script>
  <script src="main.js"></script>
  <script src="risk.js"></script>

</body>


</html>